### PR TITLE
refactor: move file contents memoization to [Build_system]

### PR DESCRIPTION
**Pull request by rgrinberg**
and remove it from the rules

Continuation of the splitting of https://github.com/ocaml/dune/pull/9132

### CODE REVIEW

File: src/dune_rules/recursive_include.ml [hunk 1]
```diff
@@ -63,7 +63,7 @@ let decode ~base_term ~include_keyword ~include_allowed_in_versions ~non_sexp_be
 
 let load_included_file config path ~context =
   let open Memo.O in
-  let+ contents = Build_system.read_file (Path.build path) ~f:Io.read_file in
+  let+ contents = Build_system.with_file (Path.build path) ~f:Io.read_file in
```
Comment (snowleopard) [inline src/dune_rules/recursive_include.ml:66]: Wait, why not use the memoized version here?

File: src/dune_rules/expander.ml [hunk 2]
```diff
@@ -265,7 +265,7 @@ let expand_read_macro ~dir ~source s ~read ~pack =
   let read =
     let open Memo.O in
     let+ x =
-      Build_system.read_file path ~f:(fun a -> Async.async (fun () -> read a))
+      Build_system.with_file path ~f:(fun a -> Async.async (fun () -> read a))
```
Comment (snowleopard) [inline src/dune_rules/expander.ml:268]: Not sure what `read` is, but maybe worth memoizing it.

**Review by snowleopard — APPROVED** [review]
Thanks! I left a couple of suggestions.

File: src/dune_rules/recursive_include.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/recursive_include.ml]: Ah yeah, good call. I memoized all the reads now.

**Pull request merged**